pkg/module/fetcher.go

```diff
@@ -0,0 +1,10 @@
+package module
+
+// Fetcher fetches module from an upstream source
+type Fetcher interface {
```

Comment:
not sure I understand correctly. would the `DownloadProtocol` interface replace both the `Fetcher` and `Ref`? If I misunderstood that, then everything below is totally worthless 😆. Otherwise, read on!

I'm asking because I like the separation between fetching and reading. The separation is obviously useful for disk based fetch/read scenarios (i.e. the thing gets put on disk once and read N times), but also could be useful if we wanted to somehow read directly into a CDN or somewhere else (if we had functionality to do that in the future)

From a super subjective readability perspective, fetching, then getting a reference to a place that the stuff was fetched, and then being able to read from that reference makes sense to me.

Reply:
oh, I see! I love the idea about wrapping. would it be ok if we see how it goes with using the `Fetcher` and `Ref` in #319 before we implement the wrapper? I'm asking because I bet it'll help us build the abstraction right

Reply:
sounds good to me 👍